### DISCUSSION TITLE
Repeater Component based on array

### DISCUSSION THREAD

**Discussion by 100wunder** [Budibase/budibase#9783] [Help]
Hi community, 

I want to render form components based on an array of objects coming from a MongoDB.

Here is an example JSON return:
```
{
  "pages": [
    {
      "title": "Home",
      "fields": [
        {
          "label": "Headline",
          "type": "text"
        },
        {
          "label": "Short Description",
          "type": "textarea"
        }
      ]
    }
  ] 
}
```

I would like to iterate over "fields" and render form components based on type.

**Problem:**
I don't know how to create a repeater component based on an array. It only seems to work when having a separate query that returns only the "fields".

Thanks, 
Daniel

**Comment by andz-bb** [comment]
@100wunder Hi Daniel, we currently have an open issue for something similar to what you're describing here #5658. It would be worth showing your interest there so that we can prioritise it based on community feedback! 👍

**Comment by 100wunder** [comment]
Thanks @andz-bb. #5658 is not exactly what I'm currently looking for (although this may be the next step). 
Is it generally possible to create a Repeater component based on an nested array (e.g. "fields" in my example code")?

**Reply by melohagan** [reply]
Hey @100wunder 

You could try using [conditional UI](https://docs.budibase.com/docs/conditions) to hide components within a repeater that don't match the type.

For example:

> Repeater block (iterating over **fields**)
> > Headline (show if Repeater block.type === "text")
> > Paragraph (show if Repeater block.type === "textarea")

You could also put the Repeater block inside a Form component if you wanted to include form fields.

**Comment by 100wunder** [comment]
Thanks @melohagan, 
Yes, I'm already using this and it works generally fine for rendering fields based on conditions. 

However - my problem is, that the **Repeater component** needs a specific **Data Provider**. It does not iterate over a nested array ("fields") of a Data Source.

Or do I just get this wrong?

**Comment by melohagan** [comment]
Hey @100wunder 

This is because there are two arrays, you would need two repeaters. What would probably be simpler would be to have a MongoDB query that can return the fields for a single page.

So you would have a binding in your query for the page ID or title. Then your response would be:
```js
[
        {
          "label": "Headline",
          "type": "text"
        },
        {
          "label": "Short Description",
          "type": "textarea"
        }
      ]
```

And you schema would be `| label | type |`

More on query bindings: https://docs.budibase.com/docs/custom-queries
